packages/adminrouter/extra/src/test-harness/modules/runner/common.py

```diff
@@ -256,38 +244,99 @@ class LogCatcher:
-        """A main loop where event are demultiplexed
-
-        The purpose of this function is to monitor all registered file
-        descriptors and in case when new data is available - hand of
-        taking care of it to LogWriter instance that is responsible
-        for given file descriptor.
-
-        poll() call generally seems to be easier to use and better fits
-        our use case than e.g. plain select() (no need for global lock
-        while updating FD lists):
+    _add_writers_queue = None
+
+    def _event_processing(self):
+        """The main loop where all LogCatcher operations are handled
+
+        This is the main loop of the LogCatcher instance. Its purpose is to
+        monitor all registered file descriptors and in case when new data is
+        available - hand of taking care of it to LogWriter instance that is
+        responsible for given file descriptor. On top of that it also
+        synchronizes adding and removing file descriptors to/from the set of
+        the ones that are monitored for input.
+
+        Due to the limitations of `select.pool()` call, we can only operate on
+        file descriptors here - integers provided by operating system. The
+        `select.pool()` call returns a list of file descriptors which are
+        waiting for servicing. This implies the use of a python dict
+        `self._writers` which maps file descriptors to `LogWriter` objects.
+        The operating system can close a file descriptor and open a new one
+        with the same number so adding new writers to the `self._writers` needs
+        to be synchronised with with the `select.pool()` call. The
+        synchronisation relies on two simple conditions:
+         a) all additions and removals to/from `self._writers` dict happen in
+            a single thread, communication with the thread is done using
+            thread-safe queue.
+         b) before any new LogWriter object is added to the `self._writers` dict,
+            we first check the `select.pool()` call to see if any tracked file
+            descriptors were closed and if they were - remove corresponding
+            LogWriter object from `self._writers`.
+
+        Bulletpoint b) prevents a race condition when some thread closed a file
+        descriptor and immediately after that another one opened a new FD and
+        submitted it to LogCatcher for monitoring. LogCatcher in this case will
+        always gracefully remove LogWriter object related to the old FD before
+        adding the new LogWriter in its place. Bulletpoint a) guarantees that
+        all these operations will be done in predefined order.
+
+         poll() call generally seems to be easier to use and better fits
+         our use case than e.g. plain select() (no need for global lock
+         while updating FD lists):
         * http://stackoverflow.com/a/25249958
         * http://www.greenend.org.uk/rjk/tech/poll.html
         """
         while True:
```

Comment:
I'm really surprised that we need to build our own polling in a language as high-level as python. Is there no simpler scheme? I would think that there would be some very elegant callback-based or futures-based library that would make this function super simple.

Reply:
Using `pool()` seemed simple at the very beginning of test harness, but in time the code grew, features were added, bugs were fixed and now it is the way it is :)

I can definitely look into `asyncio`&friends next time I touch this part of the code!